### PR TITLE
AsynchronousTest.testCancelGroupWS randomly fails

### LINKED ISSUE #19764: AsynchronousTest.testCancelGroupWS randomly fails.
### Expected behavior

Test always passes

### Actual behavior

Test randomly fails.

### Platforms affected

Probably all

### PR DESCRIPTION
**Pull request by mantid-roman**
Added synchronisation using condition variables.

Fixes #19764.

*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by martyngigg — COMMENTED** [review]
Any reason we can't use `std::condition_variable` & `std::mutex` here?

**Comment by mantid-roman** [comment]
Will it work with Poco threading?

**Comment by martyngigg** [comment]
Yep. The Poco thread classes are just wrappers around the native threading libraries the same as the `std` variants - I don't think it should be a problem.

**Comment by mantid-roman** [comment]
OK. I'll change.

**Comment by martyngigg** [comment]
Thanks. I think it's good to use `std` stuff where we can. We should probably think about switching the main algorithm threading to `std::thread` too but that's a different question.

**Comment by mantid-roman** [comment]
@martyngigg I was wrong, the times are the same.

**Comment by AntonPiccardoSelg** [comment]
Does this not lead to a potential condition variable deadlock? 
If the test thread acquires the mutex first, it can go all the way down to the point where it waits on cond2. Then the AsyncAlgorithm thread can get the mutex and waits on cond1. But since the test thread is already past the point where it notifies the threads waiting on cond1, AsyncAlgorithm will keep waiting on cond1, which in turn will leave the test thread waiting on cond2. Am I missing something?

File: Framework/API/test/AsynchronousTest.h [hunk 1]
```diff
@@ -143,12 +156,18 @@ class AsynchronousTest : public CxxTest::TestSuite {
   void testCancelGroupWS() {
     WorkspaceGroup_sptr groupWS = makeGroupWorkspace();
     AsyncAlgorithm alg;
+    alg.synchronise = true;
```
Comment (martyngigg) [inline Framework/API/test/AsynchronousTest.h]: Is this correct? `synchronise` is set to true on the parent object but not on the child algorithms so I don't think the `if` statement above will ever fire?

**Review by martyngigg — CHANGES_REQUESTED** [review]
I agree with Anton here. If the threads were slow to start up then the main thread will block indefinitely on the second condition.

File: Framework/API/test/AsynchronousTest.h [hunk 2]
```diff
@@ -11,12 +11,21 @@
 #include <Poco/NObserver.h>
 #include <Poco/Thread.h>
 
+#include <condition_variable>
+#include <mutex>
+
 using namespace Mantid::Kernel;
 using namespace Mantid::API;
 using namespace std;
```
Comment (martyngigg) [inline Framework/API/test/AsynchronousTest.h]: Can remove `using namespace std`? We shouldn't encourage its use in headers, even in tests.

**Review by martyngigg — COMMENTED** [review]
This looks fine to me now from a functional perspective.

I'm not too fond of the global variables though, even though they are in an an anonymous namespace. Having them here could encourage others to use condition variables in this manner. Can we move them to private variables of the test along with the whole class definition of  `AsyncAlgorithm`?

File: Framework/API/test/AsynchronousTest.h (hunk 2 above)
Comment (mantid-roman) [inline Framework/API/test/AsynchronousTest.h]: But the variables will be static members?

Comment (martyngigg) [inline Framework/API/test/AsynchronousTest.h]: Do they need to be static?

Comment (mantid-roman) [inline Framework/API/test/AsynchronousTest.h]: They have to be passed to the algorithms

Comment (martyngigg) [inline Framework/API/test/AsynchronousTest.h]: Oh yes, of course. :man_facepalming:

That's okay. At least they are taken out of the global scope and won't clash with other variables in other test headers (The final test binary includes all of the headers in a single file).

Comment (mantid-roman) [inline Framework/API/test/AsynchronousTest.h]: We cannot use `DECLARE_ALGORITHM` with a nested class

Comment (martyngigg) [inline Framework/API/test/AsynchronousTest.h]: Why do we need to use `DECLARE_ALGORITHM` at all if we are instantiating the object directly?

Comment (mantid-roman) [inline Framework/API/test/AsynchronousTest.h]: `AlgorithmFactory` is used to clone algorithms to process groups.

Comment (martyngigg) [inline Framework/API/test/AsynchronousTest.h]: Okay. Let's leave it as it is but can we capitalise the names to follow the standard for static variables and still remove the `using namespace std`.

**Review by martyngigg — APPROVED** [review]
I'm happy with these changes now.

**Comment by martyngigg** [comment]
:shipit:

**Pull request merged**